src/DistBCSR.hpp

```diff
@@ -0,0 +1,1192 @@
+      MPI_Fint fcomm = MPI_Comm_c2f(dbcsr_group);
+    //if (verbose)
+    //  c_dbcsr_finalize_lib_aux(&fcomm);
+    //else
+        c_dbcsr_finalize_lib_aux_silent(&fcomm);
+      MPI_Comm_free(&dbcsr_group);
+    };
+
+};
+
+/*
+ * Wrapper-class for dbcsr
+ */
+class DistBCSR {
+
+  private:
+    DBCSR_Environment* dbcsr_env;
+
+    dm_dbcsr dbcsr_matrix;
+    std::string mname;
+    size_t nrow;
+    size_t ncol;
+    std::vector<int> row_dims;
+    std::vector<int> col_dims;
+    double dbcsr_thresh;
+
+  public:
+    DistBCSR(){
+      dbcsr_matrix = nullptr;
+      this->free();
+    };
+
+    DistBCSR(DistBCSR& ref){
+      mname = ref.mname;
+      dbcsr_env = ref.dbcsr_env;
+      nrow = ref.nrow;
+      ncol = ref.ncol;
+      row_dims = ref.row_dims;
+      col_dims = ref.col_dims;
+      dbcsr_thresh = ref.dbcsr_thresh;
+
+      dbcsr_matrix = nullptr;
+      c_dbcsr_create_new_d(&dbcsr_matrix, mname.data(), dbcsr_env->dbcsr_dist, 'N',
+                            row_dims.data(), (int)row_dims.size(),
+                            col_dims.data(), (int)col_dims.size());
+      c_dbcsr_finalize(dbcsr_matrix);
+      this->copy(ref);
+    };
+
+    DistBCSR(const DistBCSR& ref){ ///< Constructor
+      assert(ref.dbcsr_env != nullptr);
+      mname = ref.mname;
+      mname += ":copy";
+      dbcsr_env = ref.dbcsr_env;
+      nrow = ref.nrow;
+      ncol = ref.ncol;
+      row_dims = ref.row_dims;
+      col_dims = ref.col_dims;
+      dbcsr_thresh = ref.dbcsr_thresh;
+
+      dbcsr_matrix = nullptr;
+      c_dbcsr_create_new_d(&dbcsr_matrix, mname.data(), dbcsr_env->dbcsr_dist, 'N',
+                            row_dims.data(), (int)row_dims.size(),
+                            col_dims.data(), (int)col_dims.size());
+      c_dbcsr_finalize(dbcsr_matrix);
+      this->copy(ref);
+    };
+
+    DistBCSR(DistBCSR&& ref){ ///< move Constructor
+      assert(ref.dbcsr_env != nullptr);
+      mname = ref.mname;
+      mname += ":copy";
+      dbcsr_env = ref.dbcsr_env;
+      nrow = ref.nrow;
+      ncol = ref.ncol;
+      row_dims = ref.row_dims;
+      col_dims = ref.col_dims;
+      dbcsr_thresh = ref.dbcsr_thresh;
+
+      dbcsr_matrix = nullptr;
+      c_dbcsr_create_new_d(&dbcsr_matrix, mname.data(), dbcsr_env->dbcsr_dist, 'N',
+                            row_dims.data(), (int)row_dims.size(),
+                            col_dims.data(), (int)col_dims.size());
+      c_dbcsr_finalize(dbcsr_matrix);
+      this->copy(ref);
+      // free ref...
+      ref.free();
+    };
+
+
+    DistBCSR(size_t nrow_in, size_t ncol_in, std::vector<int>& row_dims_in, std::vector<int>& col_dims_in, DBCSR_Environment* dbcsr_env_in,
+             double thr_in=0.e0, const std::string& mname_in="default matrix name"){
+      this->init(nrow_in,ncol_in,row_dims_in,col_dims_in,dbcsr_env_in,thr_in,false,mname_in);
+    };
+
+    DistBCSR(size_t ldim, std::vector<int>& dims_in, DBCSR_Environment* dbcsr_env_in, double thr_in=0.e0,
```

Comment:
an observer note: 

there are some places where `const` values are marked as such, but there are also a lot of places where they aren't. 

I bet here `dims_in`, `thr_in` and `add_zero_diag` are all `const`. I don't know what `ldim` means, but most likely it's also const.

Any reason why you pass an environment as a pointer? It would probably make sense to take this object (which wraps MPI comm and its grid?) as `std::shared_ptr<const DBCSR_Environment> dbcsr_env`. Most likely different matrices that use the same grid shall not be allowed to modify it anyhow (`const`-ness again). From the user perspective, it would be good if such things were reflected in the C++ interface.